pulsar-broker/src/main/java/org/apache/pulsar/broker/service/schema/SchemaRegistryServiceImpl.java

```diff
@@ -180,57 +182,85 @@ public CompletableFuture<List<CompletableFuture<SchemaAndMetadata>>> getAllSchem
+                                .setSchema(ByteString.copyFrom(schema.getData()))
+                                .setSchemaId(schemaId)
+                                .setUser(schema.getUser())
+                                .setDeleted(false)
+                                .setTimestamp(clock.millis())
+                                .addAllProps(toPairs(schema.getProps()))
+                                .build();
+                        long start = startTime > 0 ? startTime : this.clock.millis();
+                        CompletableFuture<SchemaVersion> persistentFuture = new CompletableFuture<>();
+                        schemaStorage
+                                .put(schemaId, info.toByteArray(), context)
+                                .thenAccept(sv -> {
+                                    if (log.isDebugEnabled()) {
+                                        log.debug("[{}] Schema is successfully added", schemaId);
+                                    }
+                                    this.stats.recordPutLatency(schemaId, this.clock.millis() - start);
+                                    persistentFuture.complete(sv);
+                                })
+                                .exceptionally(t -> {
+                                    if (t.getCause() instanceof AlreadyExistsException
+                                            || t.getCause() instanceof BadVersionException) {
+                                        // retry if put schemaLocator to zk failed caused by race condition
+                                        int retryTime = retry > 0 ? retry + 1 : 1;
+                                        if (log.isDebugEnabled()) {
+                                            log.debug("[{}] Put schema failed because of {}, retry {} times",
+                                                    t.getCause().getMessage(), schemaId, retryTime);
+                                        }
+                                        this.stats.recordPutRetry(schemaId);
+                                        putSchemaIfAbsent(schemaId, schema, strategy, start, retryTime)
+                                                .thenAccept(persistentFuture::complete)
+                                                .exceptionally(ex2 -> {
+                                                    persistentFuture.completeExceptionally(ex2);
+                                                    return null;
+                                                });
+
+                                    } else {
+                                        log.error("[{}] Put schema failed", schemaId, t.getCause());
+                                        this.stats.recordPutFailed(schemaId);
+                                        persistentFuture.completeExceptionally(t);
+                                    }
+                                    return null;
+                                });
```

Comment:
suggestion use `handleAsync` instead of `thenAccept ... exceptionally `, just like this:

```java
long start = startTime > 0 ? startTime : this.clock.millis();
return schemaStorage.put(schemaId, info.toByteArray(), context).handleAsync((sv, ex) -> {
  if (ex != null){
    return CompletableFuture.completedFuture(sv);
  }
  if (xx instanceof xxx){
    return putSchemaIfAbsent(schemaId, schema, strategy, start, retryTime + 1);
  }
  return CompletableFuture.failedFuture(ex);
});
```
this makes it easier to read.

Reply:
I'll try this, thanks for your suggestion